sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -104,6 +104,7 @@ abstract class Optimizer(catalogManager: CatalogManager)
         ReorderJoin,
         EliminateOuterJoin,
         PushDownPredicates,
+        RewriteWithExpression,
```

Comment:
it's fragile to rely on rule order, let's call `RewriteWithExpression` directly at the end of `PushDownPredicates`